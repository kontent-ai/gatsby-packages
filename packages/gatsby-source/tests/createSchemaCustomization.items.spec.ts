import { CreateSchemaCustomizationArgs } from "gatsby";
import {
  CustomPluginOptions,
} from '../src/types';
import { kontentItemsCreateSchemaCustomization } from '../src/createSchemaCustomization.items';
import { createMock } from 'ts-auto-mock';
import { Actions } from 'gatsby';
import { mocked } from 'ts-jest/dist/util/testing';
import axios from 'axios';
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

import fakeEmptyTypesResponse from './fakeEmptyTypesResponse.json';
import fakeEmptyElementsTypesResponse from './emptyElementsTypesResponse.json';

describe('kontentItemsCreateSchemaCustomization', () => {
  it('create fixed type definition', async () => {
    const api = createMock<CreateSchemaCustomizationArgs>({
      actions: createMock<Actions>({
        createTypes: jest.fn(),
      }),
    });

    const pluginConfiguration = createMock<CustomPluginOptions>({
      projectId: 'dummyProject',
      languageCodenames: ['dummyLanguage'],
    });

    mockedAxios.get.mockImplementation(url => {
      if (url.endsWith('/types')) {
        return Promise.resolve({
          data: fakeEmptyTypesResponse,
          headers: [],
        });
      } else {
        throw new Error(
          'Just `/types` endpoint should be called for schema definition.',
        );
      }
    });

    await kontentItemsCreateSchemaCustomization(api, pluginConfiguration);
    const createTypesMock: jest.Mock = mocked(api.actions.createTypes, true);
    expect(createTypesMock).toBeCalled();
    // Fix schema
    expect(createTypesMock.mock.calls[0][0]).toMatchSnapshot();
  });


  it('Empty element type does not fail', async () => {
    const api = createMock<CreateSchemaCustomizationArgs>({
      actions: createMock<Actions>({
        createTypes: jest.fn(),
      }),
      schema: {
        buildObjectType: jest.fn(),
      }
    });

    const pluginConfiguration = createMock<CustomPluginOptions>({
      projectId: 'dummyProject',
      languageCodenames: ['dummyLanguage'],
    });

    mockedAxios.get.mockImplementation(url => {
      if (url.endsWith('/types')) {
        return Promise.resolve({
          data: fakeEmptyElementsTypesResponse,
          headers: [],
        });
      } else {
        throw new Error(
          'Just `/types` endpoint should be called for schema definition.',
        );
      }
    });


    await kontentItemsCreateSchemaCustomization(api, pluginConfiguration);
    const createTypesMock = mocked(api.actions.createTypes, true);
    const buildObjectTypeMock = mocked(api.schema.buildObjectType, true);
    expect(createTypesMock).toBeCalledTimes(2);
    expect(buildObjectTypeMock).toBeCalledTimes(1);
    expect((buildObjectTypeMock.mock.calls[0][0].fields as Record<string, unknown>).system).toBeTruthy();
    expect((buildObjectTypeMock.mock.calls[0][0].fields as Record<string, unknown>).elements).toBeUndefined();
    expect(buildObjectTypeMock.mock.calls[0][0]).toMatchSnapshot();
  })
});
