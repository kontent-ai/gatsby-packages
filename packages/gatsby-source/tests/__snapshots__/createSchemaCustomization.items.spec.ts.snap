// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kontentItemsCreateSchemaCustomization Empty element type does not fail 1`] = `
{
  "fields": {
    "preferred_language": "String!",
    "system": "kontent_item_system!",
  },
  "infer": false,
  "interfaces": [
    "Node",
    "kontent_item",
  ],
  "name": "kontent_item_no_elements",
}
`;

exports[`kontentItemsCreateSchemaCustomization create fixed type definition 1`] = `
"# All __UPPER_CASE__  wildcards will be replaced by the configuration

interface kontent_item implements Node {
  id: ID!
  system: kontent_item_system!
  # elements: type is defined programmatically
  #  - it would be possible to use union type, but there are
  #    no benefits
  #    - union type is not propagated to filter
  #    - union types are not supported for querying in Gatsby
  preferred_language: String!
}

type kontent_item_system {
  codename: String!
  id: String!
  language: String!
  last_modified: Date! @dateformat
  name: String!
  type: String!
}

type kontent_item_text_element_value {
  name: String!
  type: String!
  value: String
}

type kontent_item_rich_text_element_value {
  name: String!
  type: String!
  value: String
  modular_content: [kontent_item]
    @kontent_item_language_link
  images: [kontent_item_rich_text_element_image]
  links: [kontent_item_rich_text_element_link]
}

type kontent_item_number_element_value {
  name: String!
  type: String!
  value: Float
}

type kontent_item_multiple_choice_element_value {
  name: String!
  type: String!
  value: [kontent_item_multiple_choice_element]
}

type kontent_item_date_time_element_value {
  name: String!
  type: String!
  value: Date @dateformat
}

type kontent_item_asset_element_value {
  name: String!
  type: String!
  value: [kontent_item_asset_element]
}

type kontent_item_modular_content_element_value {
  name: String!
  type: String!
  value: [kontent_item] @kontent_item_language_link
}

type kontent_item_custom_element_value {
  name: String!
  type: String!
  value: String
}

type kontent_item_taxonomy_element_value {
  name: String!
  type: String!
  taxonomy_group: String!
  value: [kontent_item_taxonomy_element]
}

type kontent_item_url_slug_element_value {
  name: String!
  type: String!
  value: String
}

type kontent_item_multiple_choice_element {
  codename: String!
  name: String!
}

type kontent_item_asset_element {
  name: String!
  description: String
  type: String!
  size: Int!
  url: String!
  width: Int
  height: Int
}

type kontent_item_taxonomy_element {
  name: String!
  codename: String!
}

type kontent_item_rich_text_element_image {
  image_id: String!
  url: String!
  description: String
  height: Int
  width: Int
}

type kontent_item_rich_text_element_link {
  link_id: String!
  codename: String!
  type: String!
  url_slug: String!
}
"
`;
